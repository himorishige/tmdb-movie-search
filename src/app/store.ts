import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import userReducer from 'src/features/user/userSlice';
import { setupListeners } from '@reduxjs/toolkit/query';
import { movieApi } from 'src/features/movie/movieQuery';

export const store = configureStore({
  reducer: {
    user: userReducer,
    [movieApi.reducerPath]: movieApi.reducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(movieApi.middleware),
});

setupListeners(store.dispatch);

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
