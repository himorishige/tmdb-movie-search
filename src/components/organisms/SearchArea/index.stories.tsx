import { Story, Meta } from '@storybook/react';
import SearchArea from '.';

export default {
  title: 'TMDB Search/Organisms/SearchArea',
  component: SearchArea,
} as Meta;

const Template: Story = (args) => <SearchArea {...args} />;

export const Default = Template.bind({});
