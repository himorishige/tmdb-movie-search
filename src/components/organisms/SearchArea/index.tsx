import { useCallback, useState, memo } from 'react';
import { useHistory, useLocation } from 'react-router-dom';
import { useForm, SubmitHandler } from 'react-hook-form';
import { Box, Button, Flex, Input } from '@chakra-ui/react';
import { useAppDispatch, useAppSelector } from 'src/app/hooks';
import { fetchSearchResults, selectLoading } from 'src/features/user/userSlice';
import { useToastMessage } from 'src/hooks/useToastMessage';
import { SearchParams } from 'src/types';

const SearchArea: React.VFC = memo(() => {
  const dispatch = useAppDispatch();
  const isLoading = useAppSelector(selectLoading);
  const history = useHistory();
  const search = useLocation().search;
  const query = new URLSearchParams(search);
  const [keyword, setKeyword] = useState(query.get('keyword') || '');
  const { showMessage } = useToastMessage();

  // Form制御
  const { register, handleSubmit } = useForm<SearchParams>();

  // 検索用ハンドラー
  const submitHandler: SubmitHandler<SearchParams> = useCallback(
    (data) => {
      // キーワードが未入力の場合
      if (!data.keyword) {
        showMessage({
          title: '検索キーワードを入力してください。',
          status: 'error',
        });
        return;
      }
      // キーワードが入力されている場合
      setKeyword(data.keyword);
      dispatch(fetchSearchResults({ keyword: data.keyword, page: 1 }));
      history.push(`/search?keyword=${data.keyword}&page=${1}`);
    },
    [dispatch, history, showMessage],
  );

  return (
    <Box>
      <form onSubmit={handleSubmit(submitHandler)}>
        <Flex>
          <Input
            {...register('keyword')}
            disabled={isLoading}
            defaultValue={keyword}
            bgColor="white"
            enterKeyHint="search"
            data-testid="input"
          />
          <Button type="submit" ml="1rem" colorScheme="blue" disabled={isLoading}>
            検索
          </Button>
        </Flex>
      </form>
    </Box>
  );
});

export default SearchArea;
