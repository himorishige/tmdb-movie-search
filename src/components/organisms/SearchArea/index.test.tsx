import renderer from 'react-test-renderer';
import SearchArea from '.';
import { store } from 'src/app/store';
import { Provider } from 'react-redux';
import { HelmetProvider } from 'react-helmet-async';
import { ChakraProvider } from '@chakra-ui/react';
import { BrowserRouter } from 'react-router-dom';

describe('Organisms/SearchArea', () => {
  test('正しくレンダリングされている', () => {
    const header = renderer
      .create(
        <HelmetProvider>
          <Provider store={store}>
            <ChakraProvider>
              <BrowserRouter>
                <SearchArea />
              </BrowserRouter>
            </ChakraProvider>
          </Provider>
        </HelmetProvider>,
      )
      .toJSON();
    expect(header).toMatchSnapshot();
  });
});
