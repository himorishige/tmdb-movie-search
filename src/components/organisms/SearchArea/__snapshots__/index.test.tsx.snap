// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Organisms/SearchArea 正しくレンダリングされている 1`] = `
Array [
  <div
    className="css-0"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="css-k008qs"
      >
        <input
          className="chakra-input css-ykzfd5"
          data-testid="input"
          defaultValue=""
          disabled={false}
          enterKeyHint="search"
          name="keyword"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
        />
        <button
          className="chakra-button css-2b4k24"
          disabled={false}
          type="submit"
        >
          検索
        </button>
      </div>
    </form>
  </div>,
  <span />,
]
`;
